api/app/auto_spatial_advisory/critical_hours.py

```diff
@@ -0,0 +1,457 @@
+            critical_hours_inputs,
+            fuel_types_by_area,
+            for_date,
+        )
+
+        if len(critical_hours_by_fuel_type) > 0:
+            critical_hours_by_zone_and_fuel_type[zone_key] = critical_hours_by_fuel_type
+
+    for zone_id, critical_hours_by_fuel_type in critical_hours_by_zone_and_fuel_type.items():
+        await save_critical_hours(db_session, zone_id, critical_hours_by_fuel_type, run_parameters_id)
+
+    return critical_hours_by_zone_and_fuel_type
+
+
+async def calculate_critical_hours(run_type: RunType, run_datetime: datetime, for_date: date):
+    """
+    Entry point for calculating critical hours.
+
+    :param run_type: The run type, either forecast or actual.
+    :param run_datetime: The date and time of the sfms run.
+    :param for_date: The date critical hours are being calculated for.
+    """
+
+    logger.info(f"Calculating critical hours for {run_type} run type on run date: {run_datetime}, for date: {for_date}")
+    perf_start = perf_counter()
+
+    async with get_async_write_session_scope() as db_session:
+        run_parameters_id = await get_run_parameters_id(db_session, RunType(run_type), run_datetime, for_date)
+        stmt = select(CriticalHours).where(CriticalHours.run_parameters == run_parameters_id)
+        exists = (await db_session.execute(stmt)).scalars().first() is not None
+
+        if exists:
+            logger.info("Critical hours already processed.")
+            return
+
+        async with ClientSession() as client_session:
+            header = await wfwx_api.get_auth_header(client_session)
+            all_stations = await get_stations_asynchronously()
+            station_codes = list(station.code for station in all_stations)
+            stations = await wfwx_api.get_wfwx_stations_from_station_codes(client_session, header, station_codes)
+            stations_by_zone: Dict[int, List[WFWXWeatherStation]] = defaultdict(list)
+            transformer = PointTransformer(4326, 3005)
+            for station in stations:
+                (x, y) = transformer.transform_coordinate(station.lat, station.long)
+                zone_id = await get_containing_zone(db_session, f"POINT({x} {y})", 3005)
+                if zone_id is not None:
+                    stations_by_zone[zone_id[0]].append(station)
+
+            await calculate_critical_hours_by_zone(db_session, header, stations_by_zone, run_parameters_id, for_date)
+
+    perf_end = perf_counter()
+    delta = perf_end - perf_start
+    logger.info(f"delta count before and after calculating critical hours: {delta}")
+
+
+#### - Helper functions for local testing of critical hours calculations.
+
+
+async def start_critical_hours():
+    async with get_async_write_session_scope() as db_session:
+        result = await get_most_recent_run_parameters(db_session, RunTypeEnum.actual, date(2024, 8, 8))
+        await calculate_critical_hours(result[0].run_type, result[0].run_datetime, result[0].for_date)
```

Comment:
My bad here. The call to `calculate_critical_hours` doesn't need to be nested within this block unless we want to pass the `db_session` into `calculate_critical_hours` and save ourselves from having to create another session in that function. Sorry.

Reply:
Oh wait, we do need to create the session in `calculate_critical_hours` as that gets called from the nats consumer.

Reply:
We need it to be nested here because of the way sqlalchemy lazily loads data when records are accessed: https://docs.sqlalchemy.org/en/20/errors.html#error-bhk3

Without nesting we get an error `Instance <RunParameters at 0x308e28190> is not bound to a Session; attribute refresh operation cannot proceed.`